### ISSUE TITLE
Nuitka doesn't support wxPython on MacOs

### ISSUE THREAD

**Issue by marxch** [Nuitka/Nuitka#799]
Hello!

Is Nuitka using virtualenv like technology? After I packaged my WxPython program with Nuitka, I ran the program with the following error prompt, which is the same as if Virtualenv were running a WxPython program.

This program needs access to the screen. Please run with a
Framework build of python, and only when you are logged in
on the main display of your Mac.

Virtualenv runs the WxPython program with python2.7 step-by-step as follows,But python3.7  is normal:

(noticeAppByPy2) mac@zqw-10 o % python -c "import wx;app=wx.App();"
This program needs access to the screen. Please run with a
Framework build of python, and only when you are logged in
on the main display of your Mac.

`import wx 
 
app = wx.App() 
window = wx.Frame(None, title = "wxPython - www.yiibai.com", size = (400,300)) 
panel = wx.Panel(window) 
label = wx.StaticText(panel, label = "Hello World", pos = (100,100)) 
window.Show(True) 
app.MainLoop()`

Now the question is, how can this error be resolved when using Nuitka to package a WxPython program?
Thank you very much!

**Comment by duolabmeng6** [comment]
请问一下如何解决 This program needs access to the screen. Please run with a
Framework build of python, and only when you are logged in
on the main display of your Mac.

**Comment by duolabmeng6** [comment]
Is there a solution

**Comment by kayhayen** [comment]
Right now I would look into this if a commercial user had this: https://nuitka.net/pages/commercial.html which is what the "funding_wanted" tag means. Otherwise, you can probably provide the information and change their code to not do what is likely a stupid check of some sorts, then I can include the solution.

**Comment by duolabmeng6** [comment]
```
# -*- coding:utf-8 -*-
import wx


class Frame(wx.Frame):
    def __init__(self):
        wx.Frame.__init__(self, None, title='', size=(400, 278),name='frame',style=541072384)
        self.panel = wx.Panel(self)
        self.Centre()
        self.btn = wx.Button(self.panel,size=(114, 38),pos=(79, 60),label='btn',name='button')

class myApp(wx.App):
    def  OnInit(self):
        self.frame = Frame()
        self.frame.Show(True)
        return True

if __name__ == '__main__':
    app = myApp()
    app.MainLoop()
```

Very simple demo

**Comment by duolabmeng6** [comment]
> 现在，如果商业用户有这个，我会调查一下：[https](https://nuitka.net/pages/commercial.html) : //nuitka.net/pages/commercial.html，这就是“funding_wanted”标签的意思。否则，您可能可以提供信息并更改他们的代码以不执行可能是某种愚蠢检查的操作，然后我可以包含解决方案。



> Right now I would look into this if a commercial user had this: https://nuitka.net/pages/commercial.html which is what the "funding_wanted" tag means. Otherwise, you can probably provide the information and change their code to not do what is likely a stupid check of some sorts, then I can include the solution.

Can you take a look wxpython

**Comment by duolabmeng6** [comment]
> 现在，如果商业用户有这个，我会调查一下：[https](https://nuitka.net/pages/commercial.html) : //nuitka.net/pages/commercial.html，这就是“funding_wanted”标签的意思。否则，您可能可以提供信息并更改他们的代码以不执行可能是某种愚蠢检查的操作，然后我可以包含解决方案。

Hello, brother Excuse me

```
pip install wxpython
```

```
# -*- coding:utf-8 -*-
import wx


class Frame(wx.Frame):
    def __init__(self):
        wx.Frame.__init__(self, None, title='', size=(400, 278),name='frame',style=541072384)
        self.panel = wx.Panel(self)
        self.Centre()
        self.btn = wx.Button(self.panel,size=(114, 38),pos=(79, 60),label='btn',name='button')

class myApp(wx.App):
    def  OnInit(self):
        self.frame = Frame()
        self.frame.Show(True)
        return True

if __name__ == '__main__':
    app = myApp()
    app.MainLoop()
```


```
nuitka --standalone --mingw64 --show-memory --show-progress --nofollow-imports --follow-import-to=need --output-dir=o testwx.py
```

Debugging environment is OK

Cannot run after compilation

```
This program needs access to the screen. Please run with a
Framework build of python, and only when you are logged in
on the main display of your Mac.
```

**Comment by kayhayen** [comment]
@duolabmeng6 please stop spamming this issue or I will have to block you

**Comment by kayhayen** [comment]
I have managed to reproduce it. Unfortunately the message is given based on a "False" return from the `wx._core` extension module, which is C code and a bit harder to analyse, but I am looking into that now.

**Comment by duolabmeng6** [comment]
> I have managed to reproduce it. Unfortunately the message is given based on a "False" return from the `wx._core` extension module, which is C code and a bit harder to analyse, but I am looking into that now.

thank; be grateful:

**Comment by kayhayen** [comment]
Interestingly it already happens in the case of accelerated mode only, i.e. standalone is not to blame at all, but there is a failure to execute the calls to the macOS UI libraries, I have found this code, which I believe is behind it:

```
// Function to test if the Display (or whatever is the platform equivallent)
// can be connected to.
bool wxPyApp::IsDisplayAvailable()
{
#ifdef __WXGTK__
    Display* display;
    display = XOpenDisplay(NULL);
    if (display == NULL)
        return false;
    XCloseDisplay(display);
    return true;
#endif

#ifdef __WXMAC__
    // This is adapted from Python's Mac/Modules/MacOS.c in the
    // MacOS_WMAvailable function.
    bool rv;
    ProcessSerialNumber psn;

    /*
    ** This is a fairly innocuous call to make if we don't have a window
    ** manager, or if we have no permission to talk to it. It will print
    ** a message on stderr, but at least it won't abort the process.
    ** It appears the function caches the result itself, and it's cheap, so
    ** no need for us to cache.
    */
#ifdef kCGNullDirectDisplay
    /* On 10.1 CGMainDisplayID() isn't available, and
    ** kCGNullDirectDisplay isn't defined.
    */
    if (CGMainDisplayID() == 0) {
        rv = false;
    } else
#endif
    {
        // Also foreground the application on the first call as a side-effect.
        // 
        // TODO: These APIs are deprecated, and will likely be gone in the 10.15 SDK
        //
        // [NSRunningApplication activateIgnoringOtherApps: YES]  ??
        // 
        if (GetCurrentProcess(&psn) < 0 || SetFrontProcess(&psn) < 0) {
            rv = false;
        } else {
            rv = true;
        }
    }
    return rv;
#endif

#ifdef __WXMSW__
    // TODO...
    return true;
#endif
}
```

This is fairly innocent code, which in case of standalone mode just returns false, but for accelerated there are signs of its attempt to resolve the framework path:

```
2021-09-06 15:40:05.631 Mini.bin[30101:158112] There was an error parsing the Info.plist for the bundle at URL <0x7f86af62cc50>: NSCocoaErrorDomain - 3840
2021-09-06 15:40:05.633 Mini.bin[30101:158112] There was an error parsing the Info.plist for the bundle at URL <0x7f86af447320>: NSCocoaErrorDomain - 3840
```

So basically this means much like Python itself, we will have to be a bundle. Right now, Nuitka doesn't do that yet, as single file was working for everything, and the Python framework is not enforcing usage as a framework. I tried adding '-framework cocoa' ,  and will search for more, it seems we might get away with linker flags only, or we would have to bite the bullet and implement being an application bundle and immitate the Python bundle more closely (which would be sad).

**Comment by kayhayen** [comment]
This removes wx from the equation entirely. It reproduces above C code in Python. It seems the framework loads with a simple "find_library", in Python and while it reports an error message to not locating the plist as above, there is no reason to assume the framework didn't load, the first call works, while the second to bring the process to the front does not.

```
from ctypes.util import find_library
from ctypes import cdll, c_int, pointer, Structure

app_services = cdll.LoadLibrary(find_library("ApplicationServices"))

if app_services.CGMainDisplayID() == 0:
    print("no gui")
else:
    print("gui")

print(app_services)

class ProcessSerialNumber(Structure):
    _fields_ = [("highLongOfPSN", c_int),
                ("lowLongOfPSN", c_int)]
psn = ProcessSerialNumber()
psn_p = pointer(psn)
if app_services.GetCurrentProcess(psn_p) < 0:
    print("cannot run without OS X gui process")
elif        (app_services.SetFrontProcess(psn_p) < 0):
    print("cannot set to front without OS X gui process")
else:
    print("works")

```

**Comment by kayhayen** [comment]
Fogive my failure to update. With the new option `--macos-create-app-bundle` and `--macos-disable-console` the wx works fine. This bundle mode currently only works in standalone mode, but not in onefile mode. It's required for macOS and wx to be a GUI program, which seems impossible without being a bundle.
